### PR TITLE
feat(junit): add includeRetries option to emit each retry as separate testcase

### LINKED ISSUE #29446: [Feature]: Add all retries as test runs to JUnit report
### 🚀 Feature Request

A config option for the `junit` reporter to add all retries to the output file.

### Example

_No response_

### Motivation

We're mostly using the junit reports together with [flaky detection](https://github.com/WithSecureOpenSource/flaky-tests-detection)  to monitor and detect flaky tests.

We currently have to use the `retries` config option which masks some of the flakiness, because the tests appear as success in the report.

I would love to be able to include the retries in our analysis for a more complete and precise flakiness detection.

### PR DESCRIPTION
**Pull request by vazidmansuri005**
## Summary

Adds an `includeRetries` option to the JUnit reporter that emits retry information using the [Maven Surefire JUnit XML schema](https://maven.apache.org/surefire/maven-surefire-plugin/examples/rerun-failing-tests.html):

- **Flaky tests** (eventually pass after retries): `<flakyFailure>` elements with `<stackTrace>`, no `<failure>` — test counts as **passed**
- **Permanent failures** (fail all retries): `<failure>` for the first attempt + `<rerunFailure>` elements for subsequent retries — test counts as **failed**

This follows the established schema from `maven-surefire-plugin` and is natively supported by **Jenkins** (JUnit plugin, since late 2025), **Allure**, and **Gradle** (`mergeReruns`). Tools that don't recognize these elements (GitHub Actions, GitLab CI, etc.) safely ignore them, so flaky tests still show as passed.

### Configuration

```ts
// playwright.config.ts
export default {
  retries: 2,
  reporter: [['junit', { includeRetries: true }]],
};
```

Or via environment variable:
```
PLAYWRIGHT_JUNIT_INCLUDE_RETRIES=1
```

### XML Output Examples

**Flaky test** (fails twice, passes on retry #2):
```xml
<testcase name="my test" classname="a.test.js" time="0.5">
  <flakyFailure message="expected true" type="expect.toBe">
    <stackTrace>Error: expect(received).toBe(expected)...</stackTrace>
  </flakyFailure>
  <flakyFailure message="expected true" type="expect.toBe">
    <stackTrace>Error: expect(received).toBe(expected)...</stackTrace>
  </flakyFailure>
</testcase>
<!-- tests="1" failures="0" — flaky test counts as passed -->
```

**Permanent failure** (fails all retries):
```xml
<testcase name="my test" classname="a.test.js" time="1.2">
  <failure message="expected true" type="expect.toBe">full formatted failure</failure>
  <rerunFailure message="expected true" type="expect.toBe">
    <stackTrace>Error: expect(received).toBe(expected)...</stackTrace>
  </rerunFailure>
</testcase>
<!-- tests="1" failures="1" — permanent failure counted once -->
```

### Edge Cases

| Scenario | Behavior |
|----------|----------|
| Flaky test (fails, fails, passes) | 1 testcase, 0 failures, 2 `<flakyFailure>` elements |
| Permanent failure (fails all retries) | 1 testcase, 1 failure, N-1 `<rerunFailure>` elements |
| No retries configured | Standard single testcase (unchanged) |
| `includeRetries` disabled (default) | Standard behavior (unchanged) |
| Thrown error (not assertion) | Uses `<flakyError>` / `<rerunError>` instead |

### Test Plan

- [x] Flaky test emits `<flakyFailure>` elements, counts as passed
- [x] Permanent failure emits `<failure>` + `<rerunFailure>`, counts as failed
- [x] Default behavior unchanged when `includeRetries` is not set
- [x] Works with both direct and merged (blob) reports
- [x] All existing JUnit reporter tests pass

Closes #29446

### CODE REVIEW

**Comment by vazidmansuri005** [comment]
@microsoft-github-policy-service agree

**Comment by vazidmansuri005** [comment]
@yury-s @dgozman — Would appreciate your review on this. This adds an `includeRetries` option to the JUnit reporter (requested in #29446) so that each retry attempt is emitted as a separate `<testcase>`, enabling flaky test detection tools to see the full retry history. Default behavior is unchanged.

**Comment by dgozman** [comment]
@vazidmansuri005 Thank you for the PR!

I see that you chose to turn each retry into a `<testcase>`. I've also found [this example](https://maven.apache.org/surefire/maven-surefire-plugin/examples/rerun-failing-tests.html) that instead uses `flakyFailure` and `rerunFailure`. What is your intended way to consume this information from the report? Are there any existing junit report consumers out there that understand flakiness and show it nicely in some UI?

I'd like this feature to be immediately useful with existing integrations. If one would need a custom parser for such a junit report, they might as well use the `json` report that has more information already.

**Comment by vazidmansuri005** [comment]
@dgozman Great point — I've refactored the implementation to use the [Maven Surefire JUnit XML schema](https://maven.apache.org/surefire/maven-surefire-plugin/examples/rerun-failing-tests.html) with `<flakyFailure>` and `<rerunFailure>` elements instead of separate `<testcase>` entries.

**How it works now:**
- **Flaky tests** (eventually pass): `<flakyFailure>` elements with `<stackTrace>`, no `<failure>` → test counts as passed
- **Permanent failures**: `<failure>` + `<rerunFailure>` elements → test counts as failed (once)

**Honest assessment of current tool support:**

| Tool | Parses these elements? | Shows in UI? |
|------|----------------------|--------------|
| Jenkins JUnit Plugin | Yes (since [v1380, Nov 2024](https://github.com/jenkinsci/junit-plugin/pull/765)) | Data only — accessible via `CaseResult.getFlakyFailures()` API, no UI changes yet |
| Jenkins Flaky Test Handler | Yes | Yes — flaky badge, deflake button, history charts (separate plugin) |
| Gradle Develocity | Yes (via `mergeReruns`) | Yes — FLAKY outcome badge, trend dashboard |
| Allure Report | Own retry mechanism | Has retries tab + bomb icon, but driven by `historyId`, not JUnit XML elements |
| GitHub Actions, GitLab CI, Azure DevOps | No | No — but safely ignore unknown elements |

**The immediate value even without dedicated UI:**
1. **Backwards-compatible** — tools that don't understand `flakyFailure` simply ignore it. Since flaky tests have no `<failure>` element, they correctly show as **passed**. Permanent failures still have `<failure>`, so nothing breaks.
2. **Accurate test counting** — without this, a test with retries shows `tests=1 failures=0` and you lose all retry information. With this, the XML preserves each retry's stack trace and output.
3. **Established schema** — this is exactly what Maven Surefire and Gradle produce. It's the de facto standard for representing retries in JUnit XML.
4. **Programmatic access** — Jenkins exposes it via API, and any custom JUnit XML parser (like [junitparser](https://junitparser.readthedocs.io/) for Python) can read `flakyFailure`/`rerunFailure` elements.

That said, I understand your concern about immediate usefulness. If you feel this isn't compelling enough without broader UI support, I'm open to suggestions on a better approach. The `json` reporter does have more information, but the JUnit format is what most CI/CD pipelines consume for test result aggregation, so having retry data in there (even if not all tools render it) seems valuable.

File: packages/playwright/src/reporters/junit.ts [hunk 1]
```diff
@@ -185,7 +194,76 @@ class JUnitReporter implements ReporterV2 {
     }
 
     let classification: 'failure' | 'error' | null = null;
-    if (!test.ok()) {
+
+    if (isFlaky) {
```
Comment (dgozman) [inline packages/playwright/src/reporters/junit.ts:201]: I see there is a lot of duplication here. Can we introduce some helper functions to reuse the rendering between flaky/rerun/testcase that will include stdio as well?

Comment (vazidmansuri005) [reply packages/playwright/src/reporters/junit.ts:201]: Addressed — extracted two helpers to remove the duplication:

buildSurefireRetryEntry(result, prefix): shared helper that builds <flakyFailure>/<flakyError>/<rerunFailure>/<rerunError> entries with <stackTrace>, <system-out>, <system-err> children
_addFailureEntry(test, entry): shared helper that adds the primary <failure>/<error> entry from error classification (used by both retried and standard failure paths)

**Comment by vazidmansuri005** [comment]
Hi @dgozman, the changes are pushed — extracted `buildSurefireRetryEntry` and `_addFailureEntry` helpers to eliminate the duplication across flaky/rerun/testcase paths. stdio (`system-out`/`system-err`) is now handled in the shared helper as well. Would appreciate another look when you get a chance. Thanks!

File: packages/playwright/src/reporters/junit.ts [hunk 2]
```diff
@@ -143,14 +145,21 @@ class JUnitReporter implements ReporterV2 {
...
       name: 'testcase',
       attributes: {
         // Skip root, project, file
         name: namePrefix + test.titlePath().slice(3).join(' › '),
         // filename
         classname: suiteName,
-        time: (test.results.reduce((acc, value) => acc + value.duration, 0)) / 1000
+        // For flaky tests, use the last (successful) result's duration.
+        // Otherwise, use total duration across all results.
+        time: isFlaky
```
Comment (dgozman) [inline packages/playwright/src/reporters/junit.ts:161]: When not flaky, but with includeRetries, this should be the timing of the first run.

File: packages/playwright/src/reporters/junit.ts [hunk 3]
```diff
@@ -245,40 +260,92 @@ class JUnitReporter implements ReporterV2 {
...
+  children.push({ name: 'stackTrace', text: stripAnsiEscapes(stackTrace) });
+  const resultOut = result.stdout.map(s => s.toString()).join('');
+  const resultErr = result.stderr.map(s => s.toString()).join('');
+  if (resultOut)
+    children.push({ name: 'system-out', text: resultOut });
+  if (resultErr)
+    children.push({ name: 'system-err', text: resultErr });
+  return {
+    name: elementName,
+    attributes: { message: errorInfo?.message || '', type: errorInfo?.type || 'FAILURE' },
```
Comment (dgozman) [inline packages/playwright/src/reporters/junit.ts]: Let's include `time` of the retry here as well.

File: packages/playwright/src/reporters/junit.ts (hunk 2 above)
Comment (vazidmansuri005) [reply packages/playwright/src/reporters/junit.ts:161]: Done @dgozman   — for permanent failures with `includeRetries`, the testcase `time` now uses the first result's duration instead of summing all retries.

File: packages/playwright/src/reporters/junit.ts (hunk 3 above)
Comment (vazidmansuri005) [reply packages/playwright/src/reporters/junit.ts]: Done — each `<flakyFailure>`/`<rerunFailure>` entry now includes a `time` attribute with the individual retry's duration in seconds.

**Comment by dgozman** [comment]
@vazidmansuri005 Please fix the lint failures. See [contributing guide](https://github.com/microsoft/playwright/blob/main/CONTRIBUTING.md) for more details.

**Comment by vazidmansuri005** [comment]
@dgozman Fixed — added non-null assertions for `entry.children` in `_addFailureEntry` to resolve the two TS18048 errors. Lint passes cleanly now.

**Review by dgozman — APPROVED** [review]
